framework/contracts/native/version-control/src/commands.rs

```diff
@@ -31,32 +35,95 @@ use crate::{
 };
 
 /// Add new Account to version control contract
-/// Only Factory can add Account
+/// Only Account can add itself.
 pub fn add_account(
     deps: DepsMut,
     msg_info: MessageInfo,
-    account_id: AccountId,
-    account_base: Account,
     namespace: Option<String>,
+    creator: String,
 ) -> VCResult {
     let config = CONFIG.load(deps.storage)?;
 
-    // Only Factory can add new Account
-    let is_factory = config
-        .account_factory_address
-        .map(|addr| addr == msg_info.sender)
-        .unwrap_or(false);
-    if !is_factory {
-        return Err(VCError::NotAccountFactory {});
-    }
+    // Check that sender is account
+    let contract_info = cw2::query_contract_info(&deps.querier, &msg_info.sender)?;
+    let maybe_acc_module_info = ModuleInfo::try_from(contract_info)?;
+
+    ensure!(
+        maybe_acc_module_info.id() == ACCOUNT,
+        VCError::NotAccountInfo {
+            caller_info: maybe_acc_module_info
+        }
+    );
 
-    // Check if account already exists
+    // Ensure account isn't already registered
+    let account_id = ACCOUNT_ID.query(&deps.querier, msg_info.sender.clone())?;
     ensure!(
         !ACCOUNT_ADDRESSES.has(deps.storage, &account_id),
         VCError::AccountAlreadyExists(account_id)
     );
 
-    ACCOUNT_ADDRESSES.save(deps.storage, &account_id, &account_base)?;
+    let acc_module_info = maybe_acc_module_info;
+
+    let account_code_id = REGISTERED_MODULES
+        .load(deps.storage, &acc_module_info)?
+        .unwrap_account()?;
```

Comment:
What is the benefit of checking anything except cw2 id?
User can migrate from any code id before registration and to any code id after registration, so we can't do anything against custom account contracts

Reply:
The goal is that they can't. For this we need to ensure that the migration function of the contract is properly protected.
[ORC-169: Ensure Account migration is properly protected](https://linear.app/abstract-sdk/issue/ORC-169/ensure-account-migration-is-properly-protected)

Reply:
And because registration can only happen during account instantiation, it can't be migrated from any contract, so we only need to protect migrate, ok, nice